ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/InMemory.hs

```diff
@@ -429,26 +431,34 @@ ledgerDbPush cfg ap db =
 
 -- | Push a bunch of blocks (oldest first)
 ledgerDbPushMany :: (ApplyBlock l blk, Monad m, c)
-                 => LedgerDbCfg l
+                 => (UpdateLedgerDbTraceEvent blk -> m ())
+                 -> LedgerDbCfg l
                  -> [Ap m l blk c] -> LedgerDB l -> m (LedgerDB l)
-ledgerDbPushMany = repeatedlyM . ledgerDbPush
+ledgerDbPushMany trace = repeatedlyM . pushAndTrace
+  where
+    pushAndTrace cfg ap db = do
+      trace StartedPushingBlockToTheLedgerDb
+      res <- ledgerDbPush cfg ap db
+      trace PushedBlockToTheLedgerDb
+      return res
 
 -- | Switch to a fork
 ledgerDbSwitch :: (ApplyBlock l blk, Monad m, c)
                => LedgerDbCfg l
                -> Word64          -- ^ How many blocks to roll back
+               -> (UpdateLedgerDbTraceEvent blk -> m ())
```

Comment:
In principle, this could be any action. I wonder if we want to be more specific and declare that this should actually be a tracer.

Reply:
Good question, but at least debatable. Even though I follow your line of thought and you are right, in principle this could be any hook but:
a) this could be a slipper slope to messy code in a future, because it is so easy to just "add one more thing" to a hook.
b) I've followed existing convention in our code were we pass `trace` functions as arguments

Reply:
```
data ChainSelEnv m blk = ChainSelEnv
    { lgrDB             :: LgrDB m blk
    , trace             :: TraceValidationEvent blk -> m ()
 ```
`ChainSelEnv` is using `trace` and that pollutes our design. I think I will change that to provide a `tracer` instead but in a follow up PR.

Reply:
I think it should be `Tracer`. Tracer is isomorphic to that type, and transparent, so there's no restriction gained by using either type. But `Tracer` is the tell-tale type throughout our codebase for this kind of concern.

Reply:
@nfrisby are you ok with doing this in separate PR so this one does not get polluted with to many concerns?

Reply:
I think for the the new field it should just be `Tracer`, regardless of new code, but yeah no need to change existing code in this PR.